fiat-roles/src/main/java/com/netflix/spinnaker/fiat/providers/ApplicationPrefixPermissionSource.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.fiat.providers;
+
+import com.netflix.spinnaker.fiat.model.resources.Application;
+import com.netflix.spinnaker.fiat.model.resources.Permissions;
+import javax.annotation.Nonnull;
+import lombok.Data;
+
+@Data
+public class ApplicationPrefixPermissionSource implements ResourcePermissionSource<Application> {
```

Comment:
I would make this a singleton component that accepts the configuration map of prefix -> permissions

in getPermissions iterate over the map finding matching keys (I guess all matching prefixes considering there could be overlap) and build up the resulting Permissions object from that

This could be made generic (`PrefixResourcePermissionSource<T extends Resource> implements ResourcePermissionSource<T>`) since there really isn't anything Application specific in here and I could definitely see legitimate uses for this on Account/BuildService as well. (In DefaultResourcePermissionConfig you could still just create the one for `T = Application` for now)

If you think it is more appropriate to create an instance of this per prefix mapping within a type to allow fancier ResourcePermissionProvider aggregating, it should still be done via constructor args rather than `@Data` and a setter.  In that case you'll need to update the DefaultResourcePermissionConfig to create each instance and register it with the `configurableApplicationContext.getBeanFactory().registerSingleton("name", prefixSpecificSource)` rather than returning a List.

Reply:
I apologize for the rushed implementation. I wanted to submit something quick for you to look at before I went home.

I will fix the way permissions are read from the config. But regarding having just one source that contains all the prefixes, I preferred to have one source per prefix, because, as you mentioned, it will allow to resolve them using any desired provider. If I put their resolution logic inside the source, I will be replicating the work of the `AggregatingResourcePermissionProvider`. Also, it might be desired to resolve them in different ways. For example, more specific might be desired to override less specific, which will be harder to do if I rule what resolution logic should be supported.

Reply:
agreed, I think multiple sources is the right way to go